File: antia_numerical_methods/c/xmp/round.c
/*	To round a number to specified number of digits using the given base */

#include <stdio.h>
#include <math.h>

double round(double x, int n, double b);

main()
{
	int i,j,n;
	double b,x,xr;

	for(i=0; i<99; ++i) {
		printf("type x = no. to be rounded, n = no. of digits, b = base\n");
		printf("                        (quits when n<=0)\n");
		scanf("%lf %d %lf", &x,&n,&b);

		if(n<=0) return 0;
		xr=round(x,n,b);
		printf(" %f rounded to %f \n", x,xr);
	}
	return;
}

 

/*	To round a number to N digits using base B.

	X : (input) The number to be rounded
	N : (input) The number of digits required in rounded number
	B : (input) Base of number system to be used for rounding
	The returned value of the function should be the rounded value of X

	Required routines : None
*/
#include <math.h>

double round(double x, int n, double b)

{
	int lgx, nx;
	double rx, xa, fx;

	if(x==0.0) return 0.0;

	xa=fabs(x);
	lgx=log(xa)/log(b);
	if(xa<1.0) lgx=lgx-1;
	fx=pow(b,(double) (n-lgx-1));
	nx=xa*fx+0.5;
	rx=nx/fx;
	if(x<0) rx=-rx;
	return rx;
}

File: antia_numerical_methods@Beliavsky/f77/xmp16/round.f
!     PROGRAM FOR ROUNDING A GIVEN FLOATING POINT NUMBER TO SPECIFIED
!     NO. OF DIGITS USING THE SPECIFIED BASE. ALL NUMBERS ARE PRINTED
!     OUT IN DECIMAL SYSTEM AND HENCE IT MAY BE DIFFICULT TO VERIFY THE
!     RESULT UNLESS THE BASE B=10.

      PROGRAM FROUND
      IMPLICIT REAL*16(A-H,O-Z)

51    FORMAT(1PD15.6,'  ROUNDED TO',D15.6)

100   PRINT *,'TYPE X=NO. TO BE ROUNDED,   N=NO. OF DIGITS,   B=BASE'
      PRINT *,'        (QUITS WHEN N.LE.0)'
      READ *,X,N,B
      IF(N.LE.0) STOP
      XR=ROUND(X,N,B)
      WRITE(6,51) X,XR
      GO TO 100
      END
 
!     -----------------------------------------------------
 
!	To round a number to N digits using base B.
!
!	X : (input) The number to be rounded
!	N : (input) The number of digits required in rounded number
!	B : (input) Base of number system to be used for rounding
!	The rounded value will be returned by ROUND
!
!	Required routines : None

      FUNCTION ROUND(X,N,B)
      IMPLICIT REAL*16(A-H,O-Z)

      ROUND=0.0
      IF(X.EQ.0.0) RETURN
      XA=ABS(X)
      LGX=LOG(XA)/LOG(B)
      IF(XA.LT.1.0) LGX=LGX-1
      FX=B**(N-LGX-1)
      NX=XA*FX+0.5
      XA=NX/FX
      ROUND=SIGN(XA,X)
      END
